course_discovery/apps/taxonomy_support/providers.py

```diff
@@ -87,3 +91,79 @@ def get_all_programs():  # lint-amnesty, pylint: disable=arguments-differ
                     'subtitle': program.subtitle,
                     'overview': program.overview,
                 }
+
+
+class DiscoveryXBlockMetadataProvider(XBlockMetadataProvider):
+    """
+    Discovery xblock provider.
+    """
+
+    def __init__(self):
+        """
+        Get lms client with default partner object
+        """
+        if settings.DEFAULT_PARTNER_ID:
+            partner = Partner.objects.filter(id=settings.DEFAULT_PARTNER_ID).first()
+        else:
+            partner = Partner.objects.first()
+        if partner is None:
+            raise KeyError(_("No partner object found!"))
+        self.client = LMSAPIClient(partner)
+
+    def _get_block_content(self, block_id: str) -> set:
+        """
+        Fetches block metadata i.e. `index_dictionary` using lms api and
+        returns content values as unique set.
+        """
+        block_metadata = self.client.get_blocks_metadata(block_id) or {}
+        content = block_metadata.get("index_dictionary", {}).get("content", {})
+        content_set = set(val.strip() for val in content.values() if val.strip())
+        return content_set
+
+    def _combine_text_data(self, cur_block, all_blocks):
+        """
+        Recursively combines content values in all children blocks
+        """
+        content_set = self._get_block_content(cur_block["id"])
```

Comment:
```suggestion
        content_set = self._get_block_content(cur_block.get("id", None))
```

This might raise a KeyError if id is not found.